### PR TITLE
BUG: ensure thread safety in `set_temp_(cache|config)` context managers

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
While working on #18964, I realized that these context managers where obviously not thread safe as they relying on mutations to global state.
Because they are public API, I don't think it's possible to put the genie back in the bottle now and so we must compose with this global+mutable state, and the only thing to do is hence to add locking to it.
Furthermore, because we have to allow any number of threads to call any combination of these context managers in any order, they need to share a single lock (otherwise deadlocks would be possible). In particular, the possibility of nesting more than one instance of each context imposes the use of a re-entrant lock (`RLock`).
I've added multiple tests that all fail seemingly 100% of the time on my machine so I'm confident they capture the bugs and don't think we need special jobs to run them as stress tests.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Interesting. It looks like I deadlocked the double-test job. Let's see if I can fix that quickly.

**Comment by neutrinoceros** [comment]
indeed I can easily reproduce the deadlock locally. It's hit when my new tests are run for the second time. Interestingly, running *just* these tests multiple times in a single session doesn't seem to be problematic, so what seems most likely is that some *other* test is polluting the global state somehow. I don't have a clear idea what to look for just yet but that's what I'll need to figure out next.

**Comment by neutrinoceros** [comment]
another note while I'm thinking about this problem: locking the classes isn't sufficient to provide complete safety; if two threads set a temporary cache to the *same location* concurrently, the second thread will inherit the first's cache. This is *generally* to be avoided as it creates a clear vector of attack, though I need to ponder a bit more about how to address it. I'm considering the only thing to do might be to *document* this as an unsafe pattern, and promote the use of owned uuids to set temporary directories.

**Comment by neutrinoceros** [comment]
```
uv run --extra test pytest --keep-duplicates astropy/config/tests/test_concurrency.py astropy/config/tests/test_concurrency.py
```
completes nicely

so does
```
uv run --extra test pytest --keep-duplicates astropy/config/tests/test_concurrency.py astropy/config/tests
```

but
```
uv run --extra test pytest --keep-duplicates astropy/config/tests astropy/config/tests
```
halts (deadlock)

and so does
```
uv run --extra test pytest --keep-duplicates astropy/config/tests astropy/config/tests/test_concurrency.py
```

**Comment by neutrinoceros** [comment]
In fact, I can deadlock with just
```
uv run --extra test pytest astropy/config/tests/test_configs.py astropy/config/tests/test_concurrency.py
```

which means these test modules just *happen* to pass normal jobs because their names are alphabetical sorted in the only order that doesn't cause a deadlock.
The good thing is that I wrote all these tests, so I'm uniquely well positioned to sort it out (I hope).

**Comment by neutrinoceros** [comment]
`test_set_temp_cache_resets_on_exception` is the one that's polluting the global state.

**Comment by neutrinoceros** [comment]
Problem found: I forgot to release the lock when an exception was raised. I expect clean CI this time.

File: astropy/config/paths.py [hunk 1]
```diff
@@ -134,11 +140,17 @@ def __init__(
         self._prev_path = self.__class__._temp_path
```
Comment (astrofrog) [inline astropy/config/paths.py]: This line should probably be in ``__enter__`` after the ``acquire``, otherwise if a thread reaches here while another has already executed ``__enter__``, ``self._prev_path`` will get overridden and set to ``self._path``?

File: astropy/config/paths.py [hunk 2]
```diff
@@ -292,24 +310,33 @@ def __enter__(self) -> str:
...
-        _cfgobjs.clear()
-        return super().__enter__()
+        try:
+            from .configuration import _cfgobjs
+
+            self._cfgobjs_copy = _cfgobjs.copy()
+            _cfgobjs.clear()
+            return super().__enter__()
+        except Exception:
+            _PATHS_MUTEX.release()
```
Comment (astrofrog) [inline astropy/config/paths.py]: We should probably restore ``_cfgobjs`` at this point, otherwise if an exception happens during ``super().__enter__()`` ``_cfgobjs`` will be left empty?

File: docs/changes/config/19559.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Ensure thread safety in ``set_temp_cache`` and ``set_temp_config``
+context managers
```
Comment (astrofrog) [inline docs/changes/config/19559.bugfix.rst]: Maybe worth mentioning that concurrency is now explicitly disabled inside these contexts?

File: astropy/config/paths.py [hunk 4]
```diff
@@ -134,11 +140,17 @@ def __init__(
         self._prev_path = self.__class__._temp_path
 
     def __enter__(self) -> str:
-        self.__class__._temp_path = self._path
+        _PATHS_MUTEX.acquire()
+        try:
+            self.__class__._temp_path = self._path
```
Comment (astrofrog) [inline astropy/config/paths.py:145]: Is it possible for this to raise an exception? Can we at least combine this try...except with the next one?

File: astropy/config/paths.py [hunk 5]
```diff
@@ -25,9 +26,14 @@
...
 ]
 
-
 P = ParamSpec("P")
 
+# Any number of threads may call any combination of set_temp_cache and set_temp_config,
+# in any order. Hence they need to share a single lock (otherwise deadlocks would be possible).
+# In particular, the possibility of nesting more than one instance of each context imposes
+# the use of a re-entrant lock (RLock).
+_PATHS_MUTEX = RLock()
```
Comment (astrofrog) [inline astropy/config/paths.py]: What about naming it ``_PATHS_LOCK``?

**Review by astrofrog — CHANGES_REQUESTED** [review]
This is looking good though I have a few comments below, after which we can merge

File: astropy/config/paths.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/config/paths.py]: good point

File: astropy/config/paths.py (hunk 4 above)
Comment (neutrinoceros) [reply astropy/config/paths.py:145]: it's indeed very unlikely to raise. The only failure mode I see is if `self._path` isn't defined, which is something linters and type-checkers should already be warning about. Maybe I over-corrected after my first attempt didn't have *enough* `release` hooks, but I agree this one can be dropped.

File: astropy/config/paths.py (hunk 5 above)
Comment (neutrinoceros) [reply astropy/config/paths.py]: sure. I just like "mutex" a bit more, just because it sounds slightly more ominous to me, but really, not good enough of a reason to break the convention

**Comment by neutrinoceros** [comment]
thanks a lot @astrofrog, I baked all your suggestions into a rebased commit !

File: astropy/config/paths.py [hunk 6]
```diff
@@ -294,22 +308,33 @@ def __enter__(self) -> str:
...
-        _cfgobjs.clear()
-        return super().__enter__()
+        _PATHS_LOCK.acquire()
+
+        try:
+            self._cfgobjs_copy = _cfgobjs.copy()
+            _cfgobjs.clear()
+            return super().__enter__()
+        except Exception:
+            _cfgobjs.update(self._cfgobjs_copy)
```
Comment (astrofrog) [inline astropy/config/paths.py:325]: Unlikely to happen in practice but if ``_cfgobjs.copy()`` was to ever raise an exception, ``self._cfgobjs_copy`` would be unassigned. One way around this would be to set ``self._cfgobjs_copy = None`` and checking that it isn't ``None`` here. This is just minor as unlikely, so I'll leave it up to you.

**Review by astrofrog — APPROVED** [review]
Looks good, thanks! One small remaining minor potential issue below, but I'll leave it up to you as to whether to ignore.

File: astropy/config/paths.py (hunk 6 above)
Comment (neutrinoceros) [reply astropy/config/paths.py:325]: good point. I'm going to force push again, but for ease of review, the concrete patch I'm pushing is
```patch
diff --git a/astropy/config/paths.py b/astropy/config/paths.py
index 06054e1e6d..1e313fb3c5 100644
--- a/astropy/config/paths.py
+++ b/astropy/config/paths.py
@@ -311,7 +311,14 @@ class set_temp_config(_SetTempPath):
         _PATHS_LOCK.acquire()

         try:
-            self._cfgobjs_copy = _cfgobjs.copy()
+            _cfgobjs_copy = _cfgobjs.copy()
+        except Exception:
+            _PATHS_LOCK.release()
+            raise
+
+        self._cfgobjs_copy = _cfgobjs_copy
+
+        try:
             _cfgobjs.clear()
             return super().__enter__()
         except Exception:
```
